library-best-practices/JLBP-9.md

```diff
@@ -0,0 +1,14 @@
+[JLBP-9] Ensure consumers always use a higher or equal minimum Java version
+---------------------------------------------------------------------------
+
+If a library B requiring Java 1.7 depends on a library A requiring Java 1.7, and
+library A bumps its minimum version to Java 1.8, then library B cannot pick up
+the new version of library A until it also bumps its own minimum Java
+version. This can result in a situation where the ecosystem is stuck on older
+library versions, preventing the pickup of important things like security fixes,
+and it can also lead to diamond dependency conflicts where no version works for
+all consumers. Thus, libraries should make sure their actively-maintained
```

Comment:
I don't think I agree with this statement any more. The preferable approach is to branch and maintain at least bug fixes and other mandatory changes for the Java 1.7 version.

Reply:
So branching is a code management strategy that isn't relevant to consumers. We need to talk in terms of artifacts, how those look to consumers, and how the artifacts relate to each other (e.g. naming and potential conflicts). So then, what should library owners do?

1. Create a new major version with a higher minimum Java version, retaining Java package and Maven artifact names. The old major version becomes an LTS release, and gets patches as needed, but no new major features. The dependency tree needs to settle on one of the major versions (since only one can be used at a time), which might just continue to be the old major version.
2. Create a new major version with a higher minimum Java version, renaming the Java package and Maven artifact names. The old major version becomes an LTS release, and gets patches as needed, but no new major features. The dependency tree will probably just keep using the old major version.
3. Deliver two variants of the library indefinitely having different qualifiers, with features stripped out of the one supporting a lower Java version (the Guava approach, with -jre and -android).

Reply:
Not quite. What I'm proposing is that the qualifier doesn't change, only the version numbers. For example:

Jaxen 1.x, currently 1.1.6, that is jaxen:jaxen:1.1.6 supports Java 1.4. It maintains compatibility with existing code. If we need a bug fix release, then 1.1.7, 1.1.8, maybe even 1.2.0, and so on. Maybe we never need Jaxen 1.1.7, but if we do the flexibility is there.

At the same time, Jaxen 2.0 moves forward to require Java 5 and later and make other breaking changes. This can be jaxen:jaxen:2.0.0, jaxen:jaxen:2.0.1, jaxen:jaxen:2.1.0, etc. If it makes sense, we change the group ID, artifact ID and/or package. 

Clients can upgrade on their own time, and at their own speed, without blocking the library from advancing to newer versions of Java. 

A well tested stable library should not need a lot of bug fix releases. Jaxen 1.1.6 has lasted for five years now without updates, and I'm only now beginning to think about it again because of issues with Java 9 and later.

Reply:
So it sounds like you're suggesting that library owners choose between (1) or (2) in my list.

Reply:
ping @elharo - I wanted to come to some kind of conclusion on this topic

Reply:
I like #1 and I think we should recommend that unconditionally. If the only breaking change is the minimum Java version, then there's no reason to change Maven IDs or package names.

That should happen only for major API changes.